src/main/java/org/wildfly/security/auth/realm/CacheableModifiableSecurityRealm.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.auth.realm;
+
+import org.wildfly.security.auth.server.CloseableIterator;
+import org.wildfly.security.auth.server.IdentityLocator;
+import org.wildfly.security.auth.server.ModifiableRealmIdentity;
+import org.wildfly.security.auth.server.ModifiableSecurityRealm;
+import org.wildfly.security.auth.server.RealmIdentity;
+import org.wildfly.security.auth.server.RealmUnavailableException;
+import org.wildfly.security.cache.RealmIdentityCache;
+
+/**
+ * <p>A realm which provides caching for {@link RealmIdentity} instances obtained from a {@link ModifiableSecurityRealm}..
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public final class CacheableModifiableSecurityRealm extends CacheableSecurityRealm<ModifiableSecurityRealm> implements ModifiableSecurityRealm {
```

Comment:
This is better called a "Caching...SecurityRealm" because it actually does the caching, as opposed to being capable of being cached (more on this later).